File: python/src/carlosferreyra/actions.py
"""Action handlers for menu options."""

import webbrowser

from rich.console import Console

from .config import CONFIG, is_configured_url
from .utils import animate_text, animated_spinner

console = Console()


def open_url(url: str) -> bool:
    """Open URL in default browser."""
    try:
        webbrowser.open(url)
        return True
    except Exception as e:
        console.print(f"[red]Error opening URL: {e}[/red]")
        return False


class ActionHandlers:
    """Collection of action handlers for menu options."""

    @staticmethod
    def email():
        """Open email client."""
        if not is_configured_url(CONFIG.urls.email):
            console.print("[yellow]📧 Email is currently unavailable[/yellow]")
            return

        with animated_spinner("Opening mail client..."):
            success = open_url(CONFIG.urls.email)

        if success:
            console.print("[bright_red]📧 Email client opened![/bright_red]")
            animate_text("Looking forward to hearing from you!")
        else:
            console.print("[red]❌ Failed to open email client[/red]")

    @staticmethod
    def view_resume():
        """Open resume in browser."""
        if not is_configured_url(CONFIG.urls.resume):
            console.print("[yellow]📄 Resume is currently unavailable[/yellow]")
            return

        with animated_spinner("Preparing to open resume..."):
            success = open_url(CONFIG.urls.resume)

        if success:
            console.print("[green]📥 Resume opened in your browser! 🎉[/green]")
            animate_text("Tip: You can download it directly")
        else:
            console.print("[red]❌ Failed to open resume[/red]")

    @staticmethod
    def view_portfolio():
        """Open portfolio website."""
        if not is_configured_url(CONFIG.urls.portfolio):
            console.print("[yellow]🌐 Portfolio is currently unavailable[/yellow]")
            return

        with animated_spinner("Loading portfolio..."):
            success = open_url(CONFIG.urls.portfolio)

        if success:
            console.print("[bright_cyan]🌐 Portfolio opened![/bright_cyan]")
            animate_text("Hope you enjoy exploring my work!")
        else:
            console.print("[red]❌ Failed to open portfolio[/red]")

    @staticmethod
    def view_github():
        """Open GitHub profile."""
        if not is_configured_url(CONFIG.urls.github):
            console.print("[yellow]💻 GitHub is currently unavailable[/yellow]")
            return

        with animated_spinner("Opening GitHub..."):
            success = open_url(CONFIG.urls.github)

        if success:
            console.print("[bright_green]💻 GitHub profile opened![/bright_green]")
            animate_text("Check out my latest projects!")
        else:
            console.print("[red]❌ Failed to open GitHub[/red]")

    @staticmethod
    def view_linkedin():
        """Open LinkedIn profile."""
        if not is_configured_url(CONFIG.urls.linkedin):
            console.print("[yellow]💼 LinkedIn is currently unavailable[/yellow]")
            return

        with animated_spinner("Opening LinkedIn..."):
            success = open_url(CONFIG.urls.linkedin)

        if success:
            console.print("[bright_blue]💼 LinkedIn profile opened![/bright_blue]")
            animate_text("Let's connect professionally!")
        else:
            console.print("[red]❌ Failed to open LinkedIn[/red]")

    @staticmethod
    def view_twitter():
        """Open Twitter profile."""
        if is_configured_url(CONFIG.urls.twitter):
            with animated_spinner("Opening Twitter..."):
                success = open_url(CONFIG.urls.twitter or "")

            if success:
                console.print("[bright_cyan]🐦 Twitter profile opened![/bright_cyan]")
                animate_text("Follow me for tech updates!")
            else:
                console.print("[red]❌ Failed to open Twitter[/red]")
        else:
            console.print("[yellow]Twitter profile not configured[/yellow]")


# Create action handlers instance
action_handlers = {
    "email": ActionHandlers.email,
    "view_resume": ActionHandlers.view_resume,
    "view_portfolio": ActionHandlers.view_portfolio,
    "view_github": ActionHandlers.view_github,
    "view_linkedin": ActionHandlers.view_linkedin,
    "view_twitter": ActionHandlers.view_twitter,
}

File: typescript/src/actions.ts
import chalk from 'chalk';
import gradient from 'gradient-string';
import open from 'open';
import { CONFIG, isConfiguredUrl } from './config.js';
import { animateText, createAnimatedSpinner } from './utils.js';

export const actionHandlers = {
	email: async (): Promise<void> => {
		if (!isConfiguredUrl(CONFIG.urls.email)) {
			console.log(chalk.yellow('📧 Email is currently unavailable'));
			return;
		}

		const spinner = await createAnimatedSpinner('Opening mail client...');
		await open(CONFIG.urls.email);
		spinner.success({ text: gradient.passion('📧 Email client opened!') });
		await animateText(chalk.green('Looking forward to hearing from you!'));
	},

	viewResume: async (): Promise<void> => {
		if (!isConfiguredUrl(CONFIG.urls.resume)) {
			console.log(chalk.yellow('📄 Resume is currently unavailable'));
			return;
		}

		const spinner = await createAnimatedSpinner('Preparing to open resume...');
		try {
			await open(CONFIG.urls.resume);
			spinner.success({ text: chalk.green('Resume opened in your browser! 🎉') });
			await animateText(chalk.gray('Tip: You can download it directly'));
		} catch (err) {
			spinner.error({ text: chalk.red('Failed to open resume 😢') });
			console.error(chalk.red('Error:'), (err as Error).message);
			throw err;
		}
	},

	viewPortfolio: async (): Promise<void> => {
		if (!isConfiguredUrl(CONFIG.urls.portfolio)) {
			console.log(chalk.yellow('🌐 Portfolio is currently unavailable'));
			return;
		}

		const spinner = await createAnimatedSpinner('Loading portfolio...');
		await open(CONFIG.urls.portfolio);
		spinner.success({ text: gradient.teen('🌐 Portfolio opened!') });
		await animateText(chalk.green('Hope you enjoy exploring my work!'));
	},

	viewGitHub: async (): Promise<void> => {
		if (!isConfiguredUrl(CONFIG.urls.github)) {
			console.log(chalk.yellow('💻 GitHub is currently unavailable'));
			return;
		}

		const spinner = await createAnimatedSpinner('Opening GitHub...');
		await open(CONFIG.urls.github);
		spinner.success({ text: gradient.atlas('💻 GitHub profile opened!') });
		await animateText(chalk.green('Check out my latest projects!'));
	},

	viewLinkedIn: async (): Promise<void> => {
		if (!isConfiguredUrl(CONFIG.urls.linkedin)) {
			console.log(chalk.yellow('💼 LinkedIn is currently unavailable'));
			return;
		}

		const spinner = await createAnimatedSpinner('Opening LinkedIn...');
		await open(CONFIG.urls.linkedin);
		spinner.success({ text: gradient.morning('💼 LinkedIn profile opened!') });
		await animateText(chalk.green("Let's connect professionally!"));
	},

	viewTwitter: async (): Promise<void> => {
		if (!isConfiguredUrl(CONFIG.urls.twitter)) {
			console.log(chalk.yellow('🐦 Twitter is currently unavailable'));
			return;
		}

		if (CONFIG.urls.twitter) {
			const spinner = await createAnimatedSpinner('Opening Twitter...');
			await open(CONFIG.urls.twitter);
			spinner.success({ text: gradient.cristal('🐦 Twitter profile opened!') });
			await animateText(chalk.green('Follow me for tech updates!'));
		}
	},
};
